### PR TITLE
Big deps update [AS-736]

### PR DESCRIPTION
**Pull request by zarsky-broad**
One functional change, otherwise only a couple changes due to updating React, see [here](https://reactjs.org/blog/2020/08/10/react-v17-rc.html#changes-to-event-delegation) and [here](https://reactjs.org/blog/2020/08/10/react-v17-rc.html#effect-cleanup-timing) for explanation.

Edit: I'm breaking out the markdown changes into a separate PR.

~The functional change is to markdown rendering: previously, we escaped markdown contents, which had the effect of making html tags render literally, counter to the markdown specification.~

~Now, we purify html tags properly, which means that you can use them, but things like scripts will be removed. This is safer than the old way we were handling it.~
~According to some querying of the prod database, less than 1% of workspaces have html-like code in their descriptions, so I don't expect this to effect many people, although someone out there is going to be very happy that they can now embed images and write superscripts.~

### CODE REVIEW

File: integration-tests/utils/terra-sa-utils.js [hunk 1]
```diff
@@ -3,12 +3,12 @@ const { JWT } = require('google-auth-library')
 const { getSecrets, userEmail } = require('../utils/integration-config')
 
 
-const makeAuthClient = _.memoize((subject, { client_email, private_key }) => {
+const makeAuthClient = _.memoize((subject, { client_email: email, private_key: key }) => {
```
Comment (breilly2) [inline integration-tests/utils/terra-sa-utils.js:6]: Curious... is this just a style preference? I feel like I might have chosen the style I used for a reason, but I don't remember what that reason might have been.

File: src/components/DataExplorerFrame.js [hunk 2]
```diff
@@ -11,6 +11,8 @@ const DataExplorerFrame = ({ dataset }) => {
   const { origin } = _.find({ name: dataset }, datasets)
 
   useEffect(() => {
+    const instance = elem.current
```
Comment (breilly2) [inline src/components/DataExplorerFrame.js:14]: What is the purpose of this change? (I trust that it is very intentional... I'm just trying to learn.)

File: src/components/IdleStatusMonitor.js [hunk 3]
```diff
@@ -88,10 +88,10 @@ const InactivityTimer = ({ id, timeout, countdownStart }) => {
       setLastActive(Date.now())
     }
 
-    _.forEach(event => document.addEventListener(event, updateLastActive), targetEvents)
+    _.forEach(event => document.addEventListener(event, updateLastActive, true), targetEvents)
```
Comment (breilly2) [inline src/components/IdleStatusMonitor.js:107]: Can you explain this? Docs say this last parameter is `wantsUntrusted` and is specifically for Firefox. This isn't a change I would expect to see in a deps PR.

File: src/pages/workspaces/workspace/Notebooks.js [hunk 4]
```diff
@@ -215,7 +215,7 @@ const Notebooks = _.flow(
 )(({
   apps, galaxyDataDisks, name: wsName, namespace, workspace, workspace: { accessLevel, canShare, workspace: { bucketName } },
   refreshApps, onRequesterPaysError, listView, setListView
-}) => {
+}, ref) => {
```
Comment (breilly2) [inline src/pages/workspaces/workspace/Notebooks.js:218]: What is this for?

**Review by breilly2 — COMMENTED** [review]
Initial comments on clean-up changes. Still need to go through package updates.

File: integration-tests/utils/terra-sa-utils.js (hunk 1 above)
Comment (zarsky-broad) [reply integration-tests/utils/terra-sa-utils.js:6]: The linter doesn't like when we declare variables that aren't camelcase; we don't enforce linting on the test files, but I like to keep it happy anyway, which is the reason for these changes outside the `src` folder that we lint normally.

File: src/components/DataExplorerFrame.js (hunk 2 above)
Comment (zarsky-broad) [reply src/components/DataExplorerFrame.js:14]: See here, it's from the react upgrade: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#potential-issues

File: src/components/IdleStatusMonitor.js (hunk 3 above)
Comment (zarsky-broad) [reply src/components/IdleStatusMonitor.js:107]: See here, it's from the react upgrade: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues

Comment (zarsky-broad) [reply src/components/IdleStatusMonitor.js:107]: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener the third argument is capture, the fourth would be wantsUntrusted.

File: src/pages/workspaces/workspace/Notebooks.js (hunk 4 above)
Comment (zarsky-broad) [reply src/pages/workspaces/workspace/Notebooks.js:218]: All pages in a workspace wrapper are surrounded by a `forwardRef`, so if they don't declare ref in their arguments (even if it's unused), react complains in the js console.

File: src/components/markdown.js [hunk 5]
```diff
@@ -30,5 +33,15 @@ export const newWindowLinkRenderer = (href, title, text) => {
...
 export const MarkdownEditor = props => {
-  return h(Suspense, { fallback: centeredSpinner() }, [h(SimpleMDE, props)])
+  return h(Suspense, { fallback: centeredSpinner() }, [h(SimpleMDE, _.merge({
+    options: {
+      autofocus: true,
+      renderingConfig: {
+        singleLineBreaks: false
+      },
+      previewClass: ['editor-preview', 'markdown-body'],
+      previewRender: renderMarkdown,
```
Comment (breilly2) [inline src/components/markdown.js:43]: Nice! How did you discover this? I don't see it in the docs. :(

**Review by breilly2 — APPROVED** [review]
👍   LGTM. I clicked around and tested a bunch of things, including the workspace description markdown rendering and editing.

File: src/components/markdown.js (hunk 5 above)
Comment (zarsky-broad) [reply src/components/markdown.js:43]: The docs say that [`options` is passed through](https://github.com/RIP21/react-simplemde-editor#options) to the underlying library, so I [looked through its docs](https://github.com/Ionaru/easy-markdown-editor#configuration)

Comment (breilly2) [reply src/components/markdown.js:43]: Oh, oops, of course. I was looking at the wrong docs. 🤦

File: package.json [hunk 6]
```diff
@@ -4,59 +4,59 @@
...
-    "iframe-resizer": "^4.2.11",
+    "iframe-resizer": "^4.3.2",
     "igv": "2.3.5",
-    "jszip": "^3.5.0",
-    "lodash": "^4.17.20",
-    "marked": "^0.7.0",
-    "outdated-browser-rework": "^2.10.0",
+    "jszip": "^3.6.0",
+    "lodash": "^4.17.21",
+    "marked": "^2.0.3",
```
Comment (ehanna4) [inline package.json:27]: Is this supposed to be changed? I thought it looked like you said this was going to be in another PR

Comment (zarsky-broad) [reply package.json:27]: I'm upgrading the version, but I'm not overhauling the way we parse HTML in this PR.

File: package.json [hunk 7]
```diff
@@ -4,59 +4,59 @@
...
+    "react-dropzone": "^11.3.2",
+    "react-favicon": "^0.0.21",
+    "react-focus-lock": "^2.5.0",
     "react-hyperscript-helpers": "^1.2.0",
-    "react-json-view": "^1.19.1",
-    "react-modal": "^3.11.2",
-    "react-notifications-component": "^2.4.1",
-    "react-onclickoutside": "^6.9.0",
+    "react-json-view": "^1.21.3",
+    "react-modal": "^3.13.1",
```
Comment (ehanna4) [inline package.json:42]: @Shakespeared and me made it this far in testing so far

File: package.json [hunk 8]
```diff
@@ -4,59 +4,59 @@
...
+    "qs": "^6.10.1",
+    "react": "17.0.1",
+    "react-collapse": "^5.1.0",
+    "react-dom": "npm:@hot-loader/react-dom@17.0.1",
     "react-draggable": "^4.4.3",
-    "react-dropzone": "^11.2.0",
-    "react-favicon": "^0.0.18",
-    "react-focus-lock": "^2.4.1",
+    "react-dropzone": "^11.3.2",
+    "react-favicon": "^0.0.21",
```
Comment (ehanna4) [inline package.json]: @Shakespeared and i found that the BioDataCatalyst favicon is not showing up anymore (checked on dev branch locally and it works there)

File: src/components/RuntimeManager.js [hunk 9]
```diff
@@ -226,7 +226,7 @@ export default class RuntimeManager extends PureComponent {
     } else if (runtime.status === 'Running' && prevRuntime.status === 'Updating') {
       notify('success', 'Number of workers has updated successfully.')
     }
-    if (prevApp && prevApp.status !== 'RUNNING' && app && app.status === 'RUNNING') {
+    if (prevApp?.status !== 'RUNNING' && app?.status === 'RUNNING') {
```
Comment (ehanna4) [inline src/components/RuntimeManager.js]: Something is not working with this logic. When switching between tabs if Galaxy is running a new "launch galaxy" notification pops up on every tab switch which it was not previously doing. (@Shakespeared )
Screen Shot 2021-05-10 at 3 13 09 PM

Comment (zarsky-broad) [reply src/components/RuntimeManager.js]: Good catch, I re-ordered it to be clearer.

File: package.json (hunk 8 above)
Comment (zarsky-broad) [reply package.json]: try again with this latest push?

File: src/components/data/UploadPreviewTable.js [hunk 10]
```diff
@@ -41,12 +41,14 @@ const UploadDataTable = props => {
...
-    _.flow( // eslint-disable-line lodash-fp/no-unused-result
+    const loadMetadata = _.flow(
       withErrorReporting('Error loading entity data'),
       Utils.withBusyState(setMetadataLoading)
     )(async () => {
       setEntityMetadata(await Ajax(signal).Workspaces.workspace(namespace, name).entityMetadata())
-    })()
+    })
+
+    loadMetadata()
```
Comment (ehanna4) [inline src/components/data/UploadPreviewTable.js:51]: Is this for readability or does it have a functional improvement?

Comment (zarsky-broad) [reply src/components/data/UploadPreviewTable.js:51]: It makes eslint happy, so we don't have to disable it above.

File: src/libs/browser-storage.js [hunk 11]
```diff
@@ -51,7 +51,7 @@ export const listenStatic = (storage, key, fn) => {
 export const getDynamic = (storage, key) => {
   const storageKey = `dynamic-storage/${key}`
   const data = maybeParseJSON(storage.getItem(storageKey))
-  return data && data.value
+  return data?.value
```
Comment (ehanna4) [inline src/libs/browser-storage.js:54]: We are skeptical if this will always produce the same results as the current model due to returning "undefined" rather than `data`s falsey value and then relies on anything using this function to know to handle this. Could we just leave this as is?
@Shakespeared

File: src/libs/utils.js [hunk 12]
```diff
@@ -401,9 +401,10 @@ export const useCancelable = () => {
 
 export const useCancellation = () => {
   const controller = useRef()
-  useEffect(() => {
-    return () => controller.current.abort()
-  }, [])
+  useOnMount(() => {
+    const instance = controller.current
```
Comment (ehanna4) [inline src/libs/utils.js:405]: Why would breaking this into two lines be beneficial? @Shakespeared

File: src/pages/billing/List.js [hunk 13]
```diff
@@ -113,9 +113,9 @@ const NewBillingProjectModal = ({ onSuccess, onDismiss, billingAccounts, loadAcc
     onDismiss,
     shouldCloseOnOverlayClick: false,
     title: 'Create Billing Project',
-    showCancel: !(billingAccounts && billingAccounts.length === 0),
+    showCancel: !(billingAccounts?.length === 0),
     showButtons: !!billingAccounts,
-    okButton: billingAccounts && billingAccounts.length !== 0 ?
+    okButton: billingAccounts?.length !== 0 ?
```
Comment (ehanna4) [inline src/pages/billing/List.js]: Can't do optional chaining for this like this - `billingAccounts?.length !== 0` because if billingAccounts is falsey this will be true overall which is not correct

@Shakespeared and i are now scholars when it comes to optional chaining vs short circuit evaluation

File: src/pages/billing/List.js [hunk 14]
```diff
@@ -124,14 +124,14 @@ const NewBillingProjectModal = ({ onSuccess, onDismiss, billingAccounts, loadAcc
...
-    billingAccounts && billingAccounts.length === 0 && h(Fragment, [
+    billingAccounts?.length === 0 && h(Fragment, [
       `You don't have access to any billing accounts.  `,
       h(Link, {
         href: `https://support.terra.bio/hc/en-us/articles/360026182251`,
         ...Utils.newTabLinkProps
       }, ['Learn how to create a billing account.', icon('pop-out', { size: 12, style: { marginLeft: '0.5rem' } })])
     ]),
-    billingAccounts && billingAccounts.length !== 0 && h(Fragment, [
+    billingAccounts?.length !== 0 && h(Fragment, [
```
Comment (ehanna4) [inline src/pages/billing/List.js]: Same here

File: src/pages/workspaces/List.js [hunk 15]
```diff
@@ -56,7 +56,7 @@ const WorkspaceMenuContent = ({ namespace, name, onClone, onShare, onDelete }) =
   })
 
   const canRead = workspace && Utils.canRead(workspace.accessLevel)
-  const canShare = workspace && workspace.canShare
+  const canShare = workspace?.canShare
```
Comment (ehanna4) [inline src/pages/workspaces/List.js:59]: Could we add an `!!` here to be more fool-proof that `canShare` will always be a boolean ? @Shakespeared

File: src/pages/workspaces/workspace/Dashboard.js [hunk 16]
```diff
@@ -302,7 +293,7 @@ const WorkspaceDashboard = _.flow(
             }, [icon('times', { size: 14 })])
           ])
         }, tagsList),
-        !!tagsList && tagsList.length === 0 && i(['No tags yet'])
+        tagsList?.length === 0 && i(['No tags yet'])
```
Comment (ehanna4) [inline src/pages/workspaces/workspace/Dashboard.js]: How come previously the `tagsList` was coerced into a boolean? @Shakespeared

File: src/libs/browser-storage.js (hunk 11 above)
Comment (zarsky-broad) [reply src/libs/browser-storage.js:54]: So, since `maybeParseJson` will default to `undefined`, this shouldn't be an issue; either the item is unset, in which case it's always returned `undefined`, or it's an object; there aren't any other falsy values possible here.

File: src/libs/utils.js (hunk 12 above)
Comment (zarsky-broad) [reply src/libs/utils.js:405]: See https://reactjs.org/blog/2020/08/10/react-v17-rc.html#potential-issues

File: src/pages/billing/List.js (hunk 13 above)
Comment (zarsky-broad) [reply src/pages/billing/List.js]: Ahh, I'll add some clarifying parens, the original should have been written as `billingAccounts && (billingAccounts.length !== 0)` to make that clearer

Comment (zarsky-broad) [reply src/pages/billing/List.js]: (actually I'll just change it to use `_.isEmpty`)

File: src/pages/workspaces/workspace/Dashboard.js (hunk 16 above)
Comment (zarsky-broad) [reply src/pages/workspaces/workspace/Dashboard.js]: It was to make clear that we were checking for its existence, I'll change this to `_.isEmpty`

File: package.json (hunk 8 above)
Comment (ehanna4) [reply package.json]: yep works now thanks!

File: src/pages/workspaces/List.js (hunk 15 above)
Comment (zarsky-broad) [reply src/pages/workspaces/List.js:59]: Since this is only used in two places below, and both are negating it anyway, I don't think it's worth it, personally

**Review by ehanna4 — APPROVED** [review]
Looks good - did lots of code review and testing with @Shakespeared and the final commit appeared to address our concerns.👍👍👍
